src/Storages/MergeTree/IntersectionsIndexes.h

```diff
@@ -0,0 +1,206 @@
+#pragma once
+
+#include <fmt/format.h>
+#include <Storages/MergeTree/RequestResponse.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+/// A boundary of a segment (left or right)
+struct PartToRead
+{
+    PartBlockRange range;
+    String name;
+
+    bool operator==(const PartToRead & rhs) const
+    {
+        return range == rhs.range && name == rhs.name;
+    }
+
+    bool operator<(const PartToRead & rhs) const
+    {
+        /// We allow only consecutive non-intersecting ranges
+        if (rhs.range.begin > range.begin && rhs.range.begin < range.end)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Bad ranges");
```

Comment:
Looks like it's only one kind of intersection.
Correct is: `!(end <= rhs.begin || begin >= rhs.end)`

Reply:
Or something like this

Reply:
Fixed